### PR TITLE
allow wildcard hosts for mesh gateway

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/8658. I could not find a reason why this should be restricted for mesh gateway.  We need this when the same virtual service needs to be applied for "mesh" and ingress gateway.  If there is a concern for allowing this for "mesh", I can change this to allow if there one gateway other than "mesh" (but I could not think of a case for it). We have not documented this restriction in any of our APIs.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
What happens with multiple `*` VirtualServices?

**Comment by ramaraochavali** [comment]
> What happens with multiple * VirtualServices?

You mean a single VS having multiple '*' hosts? We will have to stop it from having I think - Is that what you are asking?

**Comment by howardjohn** [comment]
No, multiple with `*`. Won't they conflict?

**Comment by ramaraochavali** [comment]
> No, multiple with `*`. Won't they conflict?

I think they wont because such a config is rejected by Envoy saying "there can not be more than one wildcard domain" - So no one would have tried it.

But if that is concern, we can validate that when we build domains and add only one - WDYT?

**Comment by ramaraochavali** [comment]
But I think that is different but related problem to this though - since we allow that already allow that for non mesh gateways now.

**Comment by howardjohn** [comment]
So right now envoy will NACK? That seems like something we should be
catching..

**Comment by ramaraochavali** [comment]
> So right now envoy will NACK? That seems like something we should be
catching..

We are deduping domains so it must filtered. But I will have to add a test to confirm. If we are good with relaxing this - I can add that test and if it sends duplicate `*` can fix it.

**Comment by ramaraochavali** [comment]
@howardjohn @hzxuzhonghu added test to validate the behaviour when duplicate wild card hosts are specified across virtual services. Found a bug when virtual host specifies a wild card, we are building domains based on matching service instead of host in VS.  PTAL.

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go [hunk 1]
```diff
@@ -142,6 +142,95 @@ func TestGenerateVirtualHostDomains(t *testing.T) {
...
+	if routeCfg == nil {
+		t.Fatalf("got nil route for %s", routeName)
+	}
+	expectedHosts := map[string][]string{
+		"allow_any":  {"*"},
+		"*.com:7442": {"*.com", "*.com:7442"},
+	}
+	expectedDestination := map[string]string{
+		"allow_any":  "PassthroughCluster",
+		"*.com:7442": "outbound|7442||test-service-one.com",
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/httproute_test.go:216]: So this is picked by oldest config?

File: pilot/pkg/networking/core/v1alpha3/route/route.go [hunk 2]
```diff
@@ -138,8 +148,12 @@ func separateVSHostsAndServices(virtualService config.Config,
 		for svcHost, svc := range serviceRegistry {
 			// *.foo.global matches *.global
 			if svcHost.Matches(hostname) {
+				// Change the host name of service to the host name in virtual service as we have to build domains based
+				// on virtual service host names.
+				svc.Hostname = hostname
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:153]: is mutation ok here? Is this not some struct that is shared beyond this function?

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:153]: looks not ok, it is a pointer and shared by all proxies

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:153]: It is safe to mutate. Because we build new service objects here https://github.com/istio/istio/blob/dbf6a9e1c123bf6abbab45683b1c6c1f7660916c/pilot/pkg/networking/core/v1alpha3/httproute.go#L220

File: pilot/pkg/networking/core/v1alpha3/httproute_test.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/httproute_test.go:216]: Yes. That is correct.

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:153]: The other case listenerPort=0, it is not reconstructed

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:153]: I have copied there also

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:153]: What if we have `*.foo.global` and `*.global`? Both will match the `svc`, and then try to mutate the Hostname?

**Comment by hzxuzhonghu** [comment]
I think we need to weigh the benefits and risks.  What is the real use case, for setting `*` in mesh gateway? If `*` is set, all the HTTP traffic will be matched by this policy. There is no once for all thing, different services need different routing policies. Not to mention the conflict possibilities increased by allowing wildcard hosts.

**Comment by ramaraochavali** [comment]
> I think we need to weigh the benefits and risks. What is the real use case, for setting * in mesh gateway? If * is set, all the HTTP traffic will be matched by this policy. There is no once for all thing, different services need different routing policies. Not to mention the conflict possibilities increased by allowing wildcard hosts.

But that is not new with `*` though. The same can happen with `*.com`, `*.net`   and more over this is a valid config from gateways. This makes route reuse in virtual services difficult.  I understand your point but exposing `*` in mesh but was wondering why is it OK in gateway but not on mesh. 

Only case where it makes sense for gateway I can think of is if a gateway exposes only one app (let us call it main app and main app routes to all other services) and it is OK to route every traffic from that gateway to the app - so `*` is a valid option as the route is picked by SNI. Is that what you are thinking as well? If gateway exposes multiple services - would not we hit the same problem?

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:153]: This is no different from what we already do today. We pick first matching service. Here we are picking first matching host in VS. So should be fine.

**Comment by ramaraochavali** [comment]
Put a hold so that we can discuss fully

**Comment by hzxuzhonghu** [comment]
I agree with you gateway may face the same problem. maybe could ask for @rshriram for more details.

File: pilot/pkg/networking/core/v1alpha3/route/route.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:153]: Handled this and added in https://github.com/istio/istio/pull/34137 as this PR looks like might need more discussion.

**Comment by ramaraochavali** [comment]
Closing this because I think it does not make sense to support this at mesh. I even think we should support *.com etc at mesh (possibly at gateway also). We can add validations later.
